jgrapht-core/src/main/java/org/jgrapht/generate/DirectedScaleFreeGraphGenerator.java

```diff
@@ -0,0 +1,376 @@
+    public void generateGraph(Graph<V, E> target, Map<String, V> resultMap)
+    {
+        Set<V> newNodesSet = new HashSet<>();
+        Set<E> newEdgesSet = new HashSet<>();
+        final int MAX_VERTEX_FAILURES = 1000;
+
+        if(targetEdges == 0 || (targetEdges < 0 && targetNodes == 0))
+            return;
+        V initV = target.addVertex();
+        newNodesSet.add(initV);
+
+        int failuresCounter = 0;
+        // grow network now
+        while (targetEdges >= 0 ? targetEdges > newEdgesSet.size()
+            : newNodesSet.size() <= targetNodes)
+        {
+
+            if (failuresCounter >= MAX_VERTEX_FAILURES) {
+                final String errMsg =
+                    failuresCounter + " successive failures more than maximum allowed number.";
+                System.err.println(errMsg);
+                throw new RuntimeException(errMsg);
+            }
+
+            V v, w, newV = null, newW = null;
+            E e;
+            float tributaries = rng.nextFloat();
+            if (tributaries <= alpha) {
+                // stop adding nodes if you will exceed the target
+                if (targetEdges < 0 && newNodesSet.size() == targetNodes)
+                    break;
+                v = newV = target.addVertex();
+                w = picAVertix(target, newNodesSet, newEdgesSet, IN_DEGREE, deltaIn);
+            } else if (tributaries <= alphaPlusBeta) {
+                v = picAVertix(target, newNodesSet, newEdgesSet, OUT_DEGREE, deltaOut);
+                w = picAVertix(target, newNodesSet, newEdgesSet, IN_DEGREE, deltaIn);
+            } else {// gamma
+                    // stop adding nodes if you will exceed the target
+                if (targetEdges < 0 && newNodesSet.size() == targetNodes)
+                    break;
+                v = picAVertix(target, newNodesSet, newEdgesSet, OUT_DEGREE, deltaOut);
+                w = newW = target.addVertex();
+            }
+
+            if (v == null || w == null) {
+                if (v == null) {
+                    target.removeVertex(newW);
+                }
+                if (w == null) {
+                    target.removeVertex(newV);
+                }
+                failuresCounter++;
+                System.err
+                    .println("WARNING: failed to pick a suitable node: Trial # " + failuresCounter);
+                continue;
+            } else {
+                failuresCounter = 0;
+            }
+
+            final GraphType graphType = target.getType();
+            // check for self loops
+            if (!graphType.isAllowingSelfLoops() && v == w) {
+                failuresCounter++;
+                continue;
+            }
+            // check for multiple parallel targetEdges
+            if (!graphType.isAllowingMultipleEdges() && target.containsEdge(v, w)) {
+                failuresCounter++;
+                continue;
+            }
+
+            e = target.addEdge(v, w);
+            if (e == null) {
```

Comment:
Same comment, remove this check. The edge is not going to be null.

Reply:
It happened with me that the edge was null because the graph did not support multiple edges while the generator did. the addEdge() method just returned null and It passed silently unnoticed until the unit test caught that the number of nodes was actually == targetNodes - 1 + null.

We have to either:
1. check the null and increase the number of failures, or
1. check the compatibility of the graph parameters and the generator parameters.

Which one do you recommend?

Reply:
You are right. In other places of the library with similar issues we use option 2, so I would prefer to just throw an exception there (IllegalArgumentException) saying that the provided graph is not compatible with multiple edges.

This might also be the case with self-loops. Is there a chance that the generator returns the same self-loop twice?

Reply:
>Is there a chance that the generator returns the same self-loop twice?
Yes. Why not?
a dublicate selfe loof is yet another self loop. If the graph allows both multiple parallel edges and self loops, then every loop is a candidate for multiple self loops.

Reply:
Ok, so could you implement option 2?

Reply:
Well, I threw exception when the parameters are not compatible, but I kept the check to ensure the ability to reuse a new vertex of a failed attempt.
Any way, the check won't harm, and better being safe than sorry.